stackslib/src/net/inv/nakamoto.rs

```diff
@@ -535,23 +541,38 @@ impl<NC: NeighborComms> NakamotoInvStateMachine<NC> {
     /// Returns the current reward cycle.
     fn update_reward_cycle_consensus_hashes(
         &mut self,
+        tip: &BlockSnapshot,
         sortdb: &SortitionDB,
     ) -> Result<u64, NetError> {
+        // check for reorg
+        let reorg = PeerNetwork::is_reorg(self.last_sort_tip.as_ref(), tip, sortdb);
+        if reorg {
+            // drop the last two reward cycles
+            test_debug!("Detected reorg! Refreshing inventory consensus hashes");
+            let highest_rc = self
+                .reward_cycle_consensus_hashes
+                .last_key_value()
+                .map(|(highest_rc, _)| *highest_rc)
+                .unwrap_or(0);
+
+            self.reward_cycle_consensus_hashes.remove(&highest_rc);
+            self.reward_cycle_consensus_hashes
+                .remove(&highest_rc.saturating_sub(1));
+        }
+
         let highest_rc = self
             .reward_cycle_consensus_hashes
             .last_key_value()
             .map(|(highest_rc, _)| *highest_rc)
             .unwrap_or(0);
 
-        let sn = SortitionDB::get_canonical_burn_chain_tip(sortdb.conn())?;
-
-        // NOTE: reward cycles start at block height mod 1, not mod 0, but
+        // NOTE: reward cycles start when (sortition_height % reward_cycle_len) == 1, not 0, but
         // .block_height_to_reward_cycle does not account for this.
         let tip_rc = sortdb
             .pox_constants
             .block_height_to_reward_cycle(
                 sortdb.first_block_height,
-                sn.block_height.saturating_sub(1),
```

Comment:
I'm really skeptical of this.

Why is it important that this distinction be made? As I've noted before, the mod 0 block does not earn rewards, but I do not see why it should be treated like a reward cycle n-1 block, and I think that's way less safe. As it stands today, computing the reward cycle of these blocks generally returns `n` _not_ `n-1`, which means that the signer set for those blocks will be `n` and _not_ `n-1`.

Reply:
This is important to inv-sync because each `GetNakamotoInv` message must contain the first consensus hash for the reward cycle (i.e. the one for sortition index 1), which proves to the receiver that the the requester has processed the same sequence of anchor blocks (and so the resulting reward cycle inventory will be meaningful to them).